contracts/programs/ocr2/src/lib.rs

```diff
@@ -600,6 +512,68 @@ pub mod ocr2 {
     }
 }
 
+fn pay_oracles_impl<'info>(
+    state: AccountLoader<'info, State>,
+    token_program: AccountInfo<'info>,
+    token_vault: AccountInfo<'info>,
+    vault_authority: AccountInfo<'info>,
+    remaining_accounts: &[AccountInfo<'info>],
+) -> Result<()> {
+    let state_id = state.key();
+    let mut state = state.load_mut()?;
+
+    require!(
+        remaining_accounts.len() == state.oracles.len(),
+        InvalidInput
+    );
+
+    let vault_nonce = state.vault_nonce;
+    let latest_round_id = state.config.latest_aggregator_round_id;
+    let billing = state.config.billing;
+
+    let payments_gjuels: Vec<(u64, CpiContext<'_, '_, '_, 'info, token::Transfer<'info>>)> = state
+        .oracles
+        .iter_mut()
+        .zip(remaining_accounts)
+        .map(|(oracle, payee)| {
+            // Ensure specified accounts match the ones inside oracles
+            require!(&oracle.payee == payee.key, InvalidInput);
+
+            let amount_gjuels = calculate_owed_payment_gjuels(&billing, oracle, latest_round_id)?;
+            // Reset reward and gas reimbursement
+            oracle.payment_gjuels = 0;
+            oracle.from_round_id = latest_round_id;
+
+            let cpi = CpiContext::new(
+                token_program.clone(),
+                token::Transfer {
+                    from: token_vault.clone(),
+                    to: payee.to_account_info(),
+                    authority: vault_authority.clone(),
+                },
+            );
+
+            Ok((amount_gjuels, cpi))
+        })
+        .collect::<Result<_>>()?;
+
+    // No account can be borrowed during CPI...
+    drop(state);
```

Comment:
yup! for some context from a internal conversation:
> You mean "No account can be borrowed during CPI..."? Both cases work, I just explicitly drop(state) so it's evident it's dropped before CPI happens. In the other case the borrow checker does this.
> 
> We can share the code for both calls, it just requires some work because the ctx type isn't the same.